federatedscope/attack/trainer/backdoor_trainer.py

```diff
@@ -0,0 +1,181 @@
+    base_trainer.ctx.target_label_ind \
+                = base_trainer.cfg.attack.target_label_ind
+    base_trainer.ctx.trigger_type = base_trainer.cfg.attack.trigger_type
+    base_trainer.ctx.label_type = base_trainer.cfg.attack.label_type
+
+    # ---- action-level plug-in -------
+
+    if base_trainer.cfg.attack.self_opt:
+
+        base_trainer.ctx.self_lr = base_trainer.cfg.attack.self_lr
+        base_trainer.ctx.self_epoch = base_trainer.cfg.attack.self_epoch
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_fit_start_init_local_opt,
+            trigger='on_fit_start',
+            insert_pos=-1)
+
+        base_trainer.register_hook_in_train(new_hook=hook_on_fit_end_reset_opt,
+                                            trigger='on_fit_end',
+                                            insert_pos=0)
+
+    scale_poisoning = base_trainer.cfg.attack.scale_poisoning
+    pgd_poisoning = base_trainer.cfg.attack.pgd_poisoning
+
+    if scale_poisoning or pgd_poisoning:
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_fit_start_init_local_model,
+            trigger='on_fit_start',
+            insert_pos=-1)
+
+    if base_trainer.cfg.attack.scale_poisoning:
+
+        base_trainer.ctx.scale_para = base_trainer.cfg.attack.scale_para
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_fit_end_scale_poisoning,
+            trigger="on_fit_end",
+            insert_pos=-1)
+
+    if base_trainer.cfg.attack.pgd_poisoning:
+
+        base_trainer.ctx.self_epoch = base_trainer.cfg.attack.self_epoch
+        base_trainer.ctx.pgd_lr = base_trainer.cfg.attack.pgd_lr
+        base_trainer.ctx.pgd_eps = base_trainer.cfg.attack.pgd_eps
+        base_trainer.ctx.batch_index = 0
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_fit_start_init_local_pgd,
+            trigger='on_fit_start',
+            insert_pos=-1)
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_batch_end_project_grad,
+            trigger='on_batch_end',
+            insert_pos=-1)
+
+        base_trainer.register_hook_in_train(
+            new_hook=hook_on_epoch_end_project_grad,
+            trigger='on_epoch_end',
+            insert_pos=-1)
+
+        base_trainer.register_hook_in_train(new_hook=hook_on_fit_end_reset_opt,
+                                            trigger='on_fit_end',
+                                            insert_pos=0)
+
+    return base_trainer
+
+
+def hook_on_fit_start_init_local_opt(ctx):
+
+    ctx.original_optimizer = ctx.optimizer
+    ctx.original_epoch = ctx["num_train_epoch"]
+    ctx["num_train_epoch"] = ctx.self_epoch
+    ctx.optimizer = torch.optim.SGD(
+        ctx.model.parameters(), lr=ctx.self_lr, momentum=0.0, weight_decay=0
+    )  # looks like adversary needs same lr to hide with others
+
+
+def hook_on_fit_end_reset_opt(ctx):
+
+    ctx.optimizer = ctx.original_optimizer
```

Comment:
In line 88: ctx.original_optimizer = ctx.optimizer and here ctx.optimizer = ctx.original_optimizer. It seems redundant?

Reply:
fixed it